homeassistant/components/zwave_js/update.py

```diff
@@ -0,0 +1,185 @@
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(
+            hass,
+            f"{DOMAIN}_{config_entry.entry_id}_add_firmware_update_entity",
+            async_add_firmware_update_entity,
+        )
+    )
+
+
+class ZWaveNodeFirmwareUpdate(UpdateEntity):
+    """Representation of a firmware update entity."""
+
+    _attr_entity_category = EntityCategory.CONFIG
+    _attr_device_class = UpdateDeviceClass.FIRMWARE
+    _attr_supported_features = (
+        UpdateEntityFeature.INSTALL | UpdateEntityFeature.RELEASE_NOTES
+    )
+    _attr_has_entity_name = True
+
+    def __init__(self, driver: Driver, node: ZwaveNode) -> None:
+        """Initialize a Z-Wave device firmware update entity."""
+        self.driver = driver
+        self.node = node
+        self.available_firmware_updates: list[FirmwareUpdateInfo] = []
+        self._latest_version_firmware: FirmwareUpdateInfo | None = None
+        self._status_unsub: Callable[[], None] | None = None
+
+        # Entity class attributes
+        self._attr_name = "Firmware"
+        self._base_unique_id = get_valueless_base_unique_id(driver, node)
+        self._attr_unique_id = f"{self._base_unique_id}.firmware_update"
+        # device may not be precreated in main handler yet
+        self._attr_device_info = DeviceInfo(
+            identifiers={get_device_id(driver, node)},
+            sw_version=node.firmware_version,
+            name=node.name or node.device_config.description or f"Node {node.node_id}",
+            model=node.device_config.label,
+            manufacturer=node.device_config.manufacturer,
+            suggested_area=node.location if node.location else None,
+        )
+
+        self._attr_installed_version = self._attr_latest_version = node.firmware_version
+
+    def _update_on_wake_up(self, _: dict[str, Any]) -> None:
+        """Update the entity when node is awake."""
+        self._status_unsub = None
+        self.hass.async_create_task(self.async_update(True))
+
+    async def async_update(self, write_state: bool = False) -> None:
+        """Update the entity."""
+        if self.node.status == NodeStatus.ASLEEP:
+            if not self._status_unsub:
+                self._status_unsub = self.node.once("wake up", self._update_on_wake_up)
+            return
+        self.available_firmware_updates = (
+            await self.driver.controller.async_get_available_firmware_updates(
+                self.node, API_KEY_FIRMWARE_UPDATE_SERVICE
+            )
+        )
+        self._async_process_available_updates(write_state)
+
+    @callback
+    def _async_process_available_updates(self, write_state: bool = True) -> None:
+        """
+        Process available firmware updates.
+
+        Sets latest version attribute and FirmwareUpdateInfo instance.
+        """
+        if self.available_firmware_updates and (
+            firmware := max(
+                self.available_firmware_updates, key=lambda x: AwesomeVersion(x.version)
```

Comment:
We no longer need to store all available firmware updates, just the latest one. Btw, is there an API to just get the latest firmware, do we always need to download the info for all available firmwares?